---
title: "Build Cross Sectional GenoPheno"
author: "Griffin Tibbitts"
date: '2023-06-26'
output: html_document
---
```{r}
rm(list = ls())
library(dplyr)
library(tidyverse)
library(gtsummary)
```

```{r}
CS_Pheno <- read.csv('/work/larylab/Griffin_Stuff/Griffin_Framingham/data/Cross_Sectional_genoPhenoData.csv')

# Filtered genotype files created from workflow described by K. Nevola
genotype_files<- list.files(path = '/work/larylab/Griffin_Stuff/Griffin_Framingham/data/TargetGenes', 
                            pattern = "_filtered_genotype.RData")

# Create single data frame with snps labeled by gene
# SNP position is given for the Ch37 human genome build
setwd('/work/larylab/Griffin_Stuff/Griffin_Framingham/data/TargetGenes')
df <- NULL
for (i in genotype_files){
  print(i)
  load(i)
  gene <- merge(tidy$gt, tidy$fix, by = "POS")
  gene$Gene <- gsub(x = i, pattern = "_.*",replacement ="")
  df <- rbind(df,gene)
}
cs_all_genes_pheno <- merge(df, CS_Pheno, by.x = "Indiv", by.y = "shareid")
```
```{r}
# Check data parameters
# Number of unique SNPs (863)
length(unique(cs_all_genes_pheno$POS)) 

# Number of people in genotype table (1591)
length(unique(cs_all_genes_pheno$Indiv)) 

# Number of People in merged table (841)
length(unique(filter(cs_all_genes_pheno, SEX == 'Female')$Indiv)) 

head(cs_all_genes_pheno)
```
```{r}
# If everything checks out, write to file. 
# Full power designates both consent groups, CS cross sectional, and FN femoral neck
save(cs_pheno_snp_pos, file = "/work/larylab/Griffin_Stuff/Griffin_Framingham/data/FULL_POWER_CS_FN_CandidateGenes.RData")
```
```{r}
head(genoPheno)
```
```{r}
# Obtain final cohort summart after merging with genotype data
cohort <- cs_all_genes_pheno %>% filter(POS == 240243682, SEX == 'Female') 

tbl_summary(cohort, include = c(BB, AGE8, BMI8, HGT8, EST8, f8cbnbmd))
```
