rm(list=ls())
library(dplyr)
options(stringsAsFactors = F)

# Pre-built file containing list of candidate genes and their ch37 build selected loci
# Modified from K. Nevola's file to include region +- 5KB from gene.
gene_list <- read.csv("/work/larylab/Griffin_Stuff/Griffin_Framingham/data/Candidate_Gene_List_v3.csv", header = T)

extract <-function(gene_list) {
  for (i in 1:nrow(gene_list)) {
    if (gene_list$Chromosome[[i]] == "10"){
      print(i)
      cmd = "vcftools --vcf /work/larylab/dbgap/data3/phg000835.v5.FHS_SHARE_imputed_HRC1.genotype-calls-vcf.c2/chr"
      cmd = paste(cmd,gene_list$Chromosome[i],"_c2.vcf", " --chr ",gene_list$Chromosome[i]," --from-bp ",gene_list$Min[i]," --to-bp ",gene_list$Max[i], sep = "")
      cmd = paste(cmd," --positions /cluster/tufts/larylabdbgap/gtibbi01/framingham/data/WellImputedPositions8.txt --recode ", "--out ",gene_list$Gene.Symbol[i],sep="")
      system(cmd)
    }else{
      print(i)
      cmd = "vcftools --gzvcf /work/larylab/dbgap/data3/phg000835.v5.FHS_SHARE_imputed_HRC1.genotype-calls-vcf.c2/chr"
      cmd = paste(cmd,gene_list$Chromosome[i],"_c2.vcf.gz", " --chr ",gene_list$Chromosome[i]," --from-bp ",gene_list$Min[i]," --to-bp ",gene_list$Max[i], sep = "")
      cmd = paste(cmd," --positions /work/larylab/Griffin_Stuff/Griffin_Framingham/data/WellImputedPositions8.txt --recode ", "--out ",gene_list$Gene.Symbol[i],sep="")
      system(cmd)
    }
  }
}

extract(gene_list = gene_list) # This will take about 3.5 hours to run
print("Done.")

i = 2
cmd = "/shared/centos7/vcftools/954e607/bin/vcftools --gzvcf /work/larylab/dbgap/data3/phg000835.v5.FHS_SHARE_imputed_HRC1.genotype-calls-vcf.c2/chr"
cmd = paste(cmd,gene_list$Chromosome[i],"_c2.vcf.gz", " --chr ",gene_list$Chromosome[i]," --from-bp ",gene_list$Min[i]," --to-bp ",gene_list$Max[i], sep = "")
cmd = paste(cmd," --positions WellImputedPositions8.txt --recode ", "--out ",gene_list$Gene.Symbol[i],sep="")
system(cmd)
