---
title: "all_merge_and _plot_v3"
author: "Griffin Tibbitts"
date: '2023-04-07'
output: html_document
---

```{r}
rm(list = ls())
```

```{r}
library(dplyr)
library(ggplot2)
library(flextable)
library(officer)
```

```{r}
# Lab BMD sample results
dxa <- read.csv('/work/larylab/Griffin_Stuff/Griffin_Mayo/data/cleaned_dxa.csv')

# Imputed BMD sample results
imp <- read.csv('/work/larylab/Griffin_Stuff/Griffin_Mayo/data/cleaned_imputed_dxa.csv')

# Remove EMmeans standard errors, they are used by other scripts
imp <- imp %>%
  select(-matches('.*pred_se'))

# Lab bone turnover marker results from blood
lab <- read.csv('/work/larylab/Griffin_Stuff/Griffin_Mayo/data/cleaned_lab.csv')

# Imputed bone turnover marker results from blood
ilab <- read.csv('/work/larylab/Griffin_Stuff/Griffin_Mayo/data/cleaned_imputed_lab.csv')

# Remove EMmeans standard errors, they are used by other scripts
ilab <- ilab %>%
  select(-matches('.*pred_se'))

# Data from Framingham offspring cohort Femoral Neck bone samples
# Drug data is not differentiated but can be assumed to be mostly atenolol
# n = 927 female participants
fram <- read.csv('/work/larylab/Griffin_Stuff/Griffin_Framingham/results/merged_with_mayo.csv')

# HDAC4 SNPs significant after bonferroni correction
corrected_fram <- read.csv('/work/larylab/Griffin_Stuff/Griffin_Framingham/results/corrected_HDAC4_merged_with_mayo.csv')

# Change headers and select to match Mayo
cleaned_fram <- fram %>%
  rename(SNP = rsid, REF = FRef, ALT = FAlt) %>%
  select(SNP, Gene, ch37, ch38, Drug, BMD_Site, Value, Std.Error, p, Source, REF, ALT) %>%
  mutate(BMD_Site = 'Femoral Neck', Drug = 'Any BB (Framingham)')

cleaned_c_fram <- corrected_fram %>%
  rename(SNP = rsid, REF = FRef, ALT = FAlt) %>%
  select(SNP, Gene, ch37, ch38, Drug, BMD_Site, Value, Std.Error, p, Source, REF, ALT) %>%
  mutate(BMD_Site = 'Femoral Neck', Drug = 'Any BB (Framingham)')
```

```{r}
# Merge imputed and non-imputed data.
# This step is done here because the imputed files are used to create 
# EMmeans plots by another script
all_dxa_merged <- rbind(dxa, imp)
all_lab_merged <- rbind(lab, ilab)
```

```{r}
parse_data_by_drug <- function(df_in, effect_estimate, standard_error, p_val, Drug_Name, Baseline) {
  
    # Calculate bounds of confidence intervals
  df_in %>%
    
    # filter for p_values significant to the specified drug and rename generic p_value column to 'p_value' and round p_value to three decimal places
    mutate(p_value = df_in[[p_val]]) %>% filter(p_value < 0.05) %>% mutate(p_value = round(p_value, 3)) %>%
    
    rename('Baseline' = Baseline, 'Value' = effect_estimate, 'Std.Error' = standard_error, 'p' = p_value) %>%

    # Make sure positions are in numeric terms.
    # Added gsub to remove hidden non-numeric characters from ch38
    mutate(ch38 = sub('.*:', '', ch38)) %>%
    mutate(ch37 = as.numeric(ch37), ch38 = as.numeric(gsub("[^0-9.-]", "", ch38))) %>%

    # Add column specifying the significant drug for merge
    mutate(Drug = Drug_Name) %>%
    
    mutate(Value = Value/Baseline)
}
# Separate BMD data by drug for which the effect was significant
atenolol <- parse_data_by_drug(all_dxa_merged, 'SNP_AvsPl_coef', 'SNP_AvsPl_se', 'SNP_AvsPl_pval', 'Atenolol', 'Mean_BMD0')
nebivolol <- parse_data_by_drug(all_dxa_merged, 'SNP_NvsPl_coef', 'SNP_NvsPl_se', 'SNP_NvsPl_pval', 'Nebivolol', 'Mean_BMD0')
propranolol_40 <- parse_data_by_drug(all_dxa_merged, 'SNP_P40vsPl_coef', 'SNP_P40vsPl_se', 'SNP_P40vsPl_pval', 'Propranolol40', 'Mean_BMD0')
propranolol_80 <- parse_data_by_drug(all_dxa_merged, 'SNP_P80vsPl_coef', 'SNP_P80vsPl_se', 'SNP_P80vsPl_pval', 'Propranolol80', 'Mean_BMD0')

# Separate Lab data by drug for which the effect was significant
atenolol_lab <- parse_data_by_drug(all_lab_merged, 'SNP_AvsPl_coef', 'SNP_AvsPl_se', 'SNP_AvsPl_pval', 'Atenolol', 'Mean_Lab0')
nebivolol_lab <- parse_data_by_drug(all_lab_merged, 'SNP_NvsPl_coef', 'SNP_NvsPl_se', 'SNP_NvsPl_pval', 'Nebivolol', 'Mean_Lab0')
propranolol_40_lab <- parse_data_by_drug(all_lab_merged, 'SNP_P40vsPl_coef', 'SNP_P40vsPl_se', 'SNP_P40vsPl_pval', 'Propranolol40', 'Mean_Lab0')
propranolol_80_lab <- parse_data_by_drug(all_lab_merged, 'SNP_P80vsPl_coef', 'SNP_P80vsPl_se', 'SNP_P80vsPl_pval', 'Propranolol80', 'Mean_Lab0')

# Re-merge into large data frames containing only results that were significant and the drug they
# were significant for.
dxa_by_drug <- bind_rows(atenolol, nebivolol, propranolol_40, propranolol_80, cleaned_fram, cleaned_c_fram)
lab_by_drug <- bind_rows(atenolol_lab, nebivolol_lab, propranolol_40_lab, propranolol_80_lab)
```

```{r}
# Certain snps from the non-imputed data use the opposite reference
# vs alternate allele from the imputed data. Manually changing the signs 
# for these snps aligns non-imputed with imputed results so that the alleles
# being counted are consistent for all snps

change_non_matches_dxa <- function(df_in){
  df_in %>% mutate(Value = case_when(SNP == "rs1042713" & Source == 'dxa' ~ as.numeric(Value) * -1,
                                     SNP == "rs1042717" & Source == 'dxa' ~ as.numeric(Value) * -1,
                                     SNP == "rs1042718" & Source == 'dxa' ~ as.numeric(Value) * -1,
                                     SNP == "rs1042719" & Source == 'dxa' ~ as.numeric(Value) * -1,
                                     SNP == "rs2050395" & Source == 'dxa' ~ as.numeric(Value) * -1,
                                     SNP == "rs12654778" & Source == 'dxa' ~ as.numeric(Value) * -1,
                                     SNP == "rs1801253" & Source == 'lab' ~ as.numeric(Value) * -1,
                                     TRUE ~ Value)) %>%
    arrange(SNP)
}

change_non_matches_lab <- function(df_in){
  df_in %>% mutate(Value = case_when(SNP == "rs1042713" & Source == 'lab' ~ as.numeric(Value) * -1,
                                     SNP == "rs1042717" & Source == 'lab' ~ as.numeric(Value) * -1,
                                     SNP == "rs1042718" & Source == 'lab' ~ as.numeric(Value) * -1,
                                     SNP == "rs1042719" & Source == 'lab' ~ as.numeric(Value) * -1,
                                     SNP == "rs2050395" & Source == 'lab' ~ as.numeric(Value) * -1,
                                     SNP == "rs12654778" & Source == 'lab' ~ as.numeric(Value) * -1,
                                     SNP == "rs1801253" & Source == 'lab' ~ as.numeric(Value) * -1,
                                     TRUE ~ Value)) %>%
    arrange(SNP)
}

matches_dxa <- change_non_matches_dxa(dxa_by_drug) 
matches_lab <- change_non_matches_lab(lab_by_drug) 
```
```{r}
# Now that the imputed and non-imputed data from Mayo is aligned,
# it needs one more sign change to ensure that the counted alleles are the same
# as Framingham. The Mayo data set is counting the number of reference alleles
# whereas Framingham is counting alternate alleles. Flipping the sign for Mayo
# gives the effect estimate of the change for each copy of the alternate allele.
flip_non_framinghams_dxa <- matches_dxa %>%
  mutate(Value = case_when(Source == 'dxa' | Source == 'Imputed' ~ Value * -1,
                           TRUE ~ Value)) %>%
  select(ch38, SNP, Gene, Drug, BMD_Site, Value, Std.Error, p, Source, REF, ALT)

flip_non_framinghams_lab <- matches_lab %>%
  mutate(Value = case_when(Source == 'dxa' | Source == 'Imputed' ~ Value * -1,
                           TRUE ~ Value)) %>%
  select(ch38, SNP, Gene, Drug, Lab, Value, Std.Error, p, Source, REF, ALT) %>%
  mutate(Value)
```

```{r}
# Create Separate Tables
# Create tables of data separated by BMD site
lumbar_spine <- flip_non_framinghams_dxa %>%
  filter(BMD_Site == 'Lumbar Spine')
femoral_neck <- flip_non_framinghams_dxa %>%
  filter(BMD_Site == 'Femoral Neck')
ultradistal_radius <- flip_non_framinghams_dxa %>%
  filter(BMD_Site == 'Ultradistal Radius')
femoral_hip <- flip_non_framinghams_dxa %>%
  filter(BMD_Site == 'Femoral Hip')
total_body <- flip_non_framinghams_dxa %>%
  filter(BMD_Site == 'Total Body')

# Data frames for each of the lab results
trap5b <- flip_non_framinghams_lab %>%
  filter(Lab == 'trap5b')
opg <- flip_non_framinghams_lab %>%
  filter(Lab == 'opg')
rankl <- flip_non_framinghams_lab %>%
  filter(Lab == 'rankl')
ctx <- flip_non_framinghams_lab %>%
  filter(Lab == 'ctx')
osteopontin <- flip_non_framinghams_lab %>%
  filter(Lab == 'osteopontin')
p1np <- flip_non_framinghams_lab %>%
  filter(Lab == 'p1np')
ocn <- flip_non_framinghams_lab %>%
  filter(Lab == 'ocn')

# Give large tables a more descriptive name for saving
filtered_dxa_data <- flip_non_framinghams_dxa
filtered_lab_data <- flip_non_framinghams_lab
```

```{r}
# Create a plot showing the effect estimate as a percentage for each SNP
# for each drug, at each site/lab type, grouped by gene.
make_plots_by_gene <- function(df_in, data_type, data_title){

  out_plot <- df_in %>% ggplot(aes(x=as.character(ch38), y=as.numeric(Value))) +
    geom_hline(yintercept = 0, linetype = 'dashed') +
    geom_point(aes(color = Drug, shape = data_type)) +  
    facet_grid(~Gene, scales = "free_x", space = 'free_x', switch = 'both') +
    labs(title = paste0(data_title, 'Pharmacogenetic Effect of SNP plus Beta Blocker vs Placebo'),
       x = 'SNP Position', 
       y = "Percent Change Per Year") +
    scale_color_manual(values = c('Atenolol' = 'orange', 'Nebivolol' = 'red', 'Propranolol40' = 'light green', 'Propranolol80' = 'dark green', 
                                'Any BB (Framingham)' = 'blue')) +
      scale_y_continuous(labels = scales::percent) +
    theme_classic(base_family = "Roboto Condensed") +
    theme(
      plot.margin = margin(0.5, 0.5, 0.5, 0.5, unit = "cm"),
      plot.title = element_text(size = 12, face = 'bold'),
      strip.text.x = element_text(face = "bold"),
      strip.placement = "outside",
      strip.background = element_blank(),
      axis.text.x = element_text(angle = 90, size = 8, vjust = .5),
      legend.key.size = unit(.15, 'cm'),
      legend.title = element_text(face = 'bold'))
  
  ggsave(paste0(data_title, '_plot.png'),
    plot = out_plot,
    device = png,
    path = '/work/larylab/Griffin_Stuff/Griffin_Mayo/results/plots/effects_plots/')
}

# Plot all significant BMD results
dxa_plot <- make_plots_by_gene(flip_non_framinghams_dxa, flip_non_framinghams_dxa$BMD_Site, 'All Significant BMD')

# Plot significant BMD results by BMD site
ls_plot <- make_plots_by_gene(lumbar_spine, lumbar_spine$BMD_Site, 'Lumbar Spine \n')
fn_plot <- make_plots_by_gene(femoral_neck, femoral_neck$BMD_Site, 'Femoral Neck \n')
ur_plot <- make_plots_by_gene(ultradistal_radius, ultradistal_radius$BMD_Site, 'Ultradistal Radius \n')
fh_plot <- make_plots_by_gene(femoral_hip, femoral_hip$BMD_Site, 'Femoral Hip \n')
tb_plot <- make_plots_by_gene(total_body, total_body$BMD_Site, 'Total Body \n')

# Plot significant lab results by type
trap5b_plot <- make_plots_by_gene(trap5b, trap5b$Lab, 'TRAP 5B (Resorption Biomarker)\n')
opg_plot <- make_plots_by_gene(opg, opg$Lab, 'Osteoprotegerin (Protects Against Resorption)\n')
rankl_plot <- make_plots_by_gene(rankl, rankl$Lab, 'RANK Ligand (Implicated in Bone Loss Pathway)\n')
ctx_plot <- make_plots_by_gene(ctx, ctx$Lab, 'ctx (Resorption Biomarker)\n ')
osteopontin_plot <- make_plots_by_gene(osteopontin, osteopontin$Lab, 'Osteopontin (Resorption Biomarker)\n')
p1np_plot <- make_plots_by_gene(p1np, p1np$Lab, 'P1NP (Resorption Biomarker)\n')
ocn_plot <- make_plots_by_gene(ocn, ocn$Lab, 'Osteocalcin (Formation Biomarker)\n')
```