---
title: "run_lmekin_longitudinal"
author: "Griffin Tibbitts"
date: '2023-04-10'
output: html_document
acknowledgement: Adapted from code written by K. Nevola and C. Lary
---

```{r}
rm(list = ls())

library(tidyverse)
library(dplyr)
library(kinship2)
library(coxme)
library(emmeans)
```

```{r}
pedigree1 <- read.csv("/work/larylab/Griffin_Stuff/Griffin_Framingham/data/share_ped_052517.csv")
load("/work/larylab/Griffin_Stuff/Griffin_Framingham/data/CandidateGenes.RData")
```

```{r}
extract_lmekin <- function(x, ...) {
  fcoef <- x$coefficients$fixed
  if (length(fcoef) > 0)  { # Not a ~1 model
    se <- sqrt(diag(x$var))
    tmp <- cbind(fcoef, se, round(fcoef/se,2),
                 signif(1 - pchisq((fcoef/ se)^2, 1), 2))
    dimnames(tmp) <- list(names(fcoef), c("Value", "Std Error", "z", "p")) 
  }
  results = data.frame(n = x$n, tmp)
}

# Run lmekin on the effect of allele count plus beta blocker use versus change 
# in BMD per year with the usual covariates as well as starting BMD (exam 6)

# Use extract_lmekin() to repackage the output as a data frame
results_lmekin <- function(data, kmat, cov){
  results <- NULL
  for (i in 1:length(data$POS)){
    print(i)
    x <- Sys.time()
    # gt_DS = genotype variable (should be 0, 1, or 2 - imperfect imputation could lead to values in between)
    # FNperyear assumes working with femoral neck, for LS make sure to change
    lmmkin1 <- lmekin(formula = as.formula(paste("FNperyear~gt_DS*BB6 +", cov, "+ (1|colnames(kmat))")), data = data$data[[i]],
                      varlist = list(kmat))
    new_results <- data.frame(extract_lmekin(lmmkin1), POS = data$POS[[i]], Gene = data$Gene[[i]])
    new_results$var <- row.names(new_results)
    results <- rbind(results, new_results)
    y<-Sys.time()
    print(y-x)
  }
  return(results)
}
```

```{r}
data<-pheno_snp_pos %>% unnest(., cols = c(data))
missing_id <- as.numeric(setdiff(data$Indiv, pedigree1$shareid))
data1 <- pheno_snp_pos$data[[1]]
missing_sex <- data1$SEX[data1$Indiv %in% missing_id]
missing_sex[missing_sex == "Female"] <- 2
missing_sex[missing_sex == "Male"] <- 1
missing_sex <- as.numeric(missing_sex)

# This line can be troublesome. If you get an error, update the pedno parameter
# to be equal to 1600:16[second row minus 1]
missing_ped <- data.frame(pedno = 1600:1678, shareid = missing_id, fshare = NA, mshare = NA, sex = missing_sex, itwin = NA)
ped_all <- rbind(pedigree1, missing_ped)
ped<-with(ped_all, pedigree(id = shareid, fshare, mshare, sex=sex, famid=pedno))
kmat <- kinship(ped)
```

```{r}
# Create sex stratified data and kmat ####
data_fem <- data %>% filter(., SEX == 'Female')
data_male <- data %>% filter(., SEX == 'Male')
length(unique(data_fem$Indiv))
```

```{r}
kmat1<-as.matrix(kmat)
ids <- colnames(kmat1) %in% data1$Indiv
pheno_kmat <- kmat1[ids,ids]

female_ids <- colnames(kmat1) %in% data_fem$Indiv
female_kmat <- kmat1[female_ids,female_ids]

male_ids <- colnames(kmat1) %in% data_male$Indiv
male_kmat <- kmat1[male_ids,male_ids]

#  Are there any Identical twins? ####
pheno_pedigree <- ped_all[ped_all$shareid %in% data1$Indiv,]
twins <- pheno_pedigree[is.na(pheno_pedigree$itwin)== F,] # Yes there are 8 individuals with the itwin marker TRUE
```

```{r}
# Important, HDAC4 is a very large gene and dramatically increases the running time
# of the model. For preliminary data and troubleshooting consider running on just
# ADRB1 and ADRB2
relevant_snp_nest <- data_fem %>% filter(., Gene %in% c("ADRB1", "ADRB2", "HDAC4")) %>% group_by(., POS, Gene) %>% nest()
```

```{r}
# Run the model and create a data frame of all of the results
# This will take some time, the console will update with run times and 
# the current position being checked and the number in the data frame
relevant_snp_results_f <- results_lmekin(data = relevant_snp_nest, kmat = female_kmat, cov = "AGE6 + HGT6 + BMI6 + EST6 + f6_7nbmd")
```
```{r}
int_relevant_snp_results_f <- relevant_snp_results_f %>% filter(., var == "gt_DS:BB6Yes") %>% filter(., p < 0.05)
int_relevant_snp_results_f

# Dividing alpha by number of tests, in this case total number of snps (820)
# The vast majority of these are HDAC4 snps
quick_correction <- relevant_snp_results_f %>% filter(., var == "gt_DS:BB6Yes") %>% filter(., p < 0.05/820)
```


```{r}
# Save the outputs. If necessary update version number
write.csv(relevant_snp_results_f, "/work/larylab/Griffin_Stuff/Griffin_Framingham/results/AllSNPs_LMEKIN_results_female_v5.csv", quote = F, row.names = F)
write.csv(int_relevant_snp_results_f, "/work/larylab/Griffin_Stuff/Griffin_Framingham/results/AllSNP_sig_interaction_LMEKIN_female_v5.csv", quote = F, row.names = F)
write.csv(quick_correction, "/work/larylab/Griffin_Stuff/Griffin_Framingham/results/AllSNP_sig_interaction_quick_correction_female_v5.csv", quote = F, row.names = F)
```

