---
title: "Build EMmeans Plots"
author: "Griffin Tibbitts"
date: '2023-05-02'
output: html_document
---

```{r}
rm(list = ls())
```

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)
```

```{r}
# Load Files
# From clean_dxa.Rmd
mayo_emmeans <- read.csv('/work/larylab/Griffin_Stuff/Griffin_Mayo/data/cleaned_imputed_dxa.csv')

mayo_snps <- read.csv('/work/larylab/Griffin_Stuff/Griffin_Mayo/results/mayo_snps_list.csv')

# From runLMER.Rmd
long_fram <- read.csv('/work/larylab/Griffin_Stuff/Griffin_Framingham/results/AllSNPs_LMER_emmeans_female.csv')
cs_fram <- read.csv('/work/larylab/Griffin_Stuff/Griffin_Framingham/results/cs_AllSNPs_LMER_emmeans_female.csv')
```

```{r}
# Emmeans plot scripting is different between the two studies. 
# Let's deal with Mayo first.

parse_data_by_drug <- function(df_in, effect_estimate, standard_error, p_val, Drug_Name) {
  
    # Calculate bounds of confidence intervals
  df_in %>%
    
    # filter for p_values significant to the specified drug and rename generic p_value column to 'p_value' and round p_value to     three decimal places
    
    mutate(p_value = df_in[[p_val]]) %>% filter(p_value < 0.05) %>% mutate(p_value = round(p_value, 3)) %>%
    
    rename('Baseline' = Mean_BMD0, 'Value' = effect_estimate, 'Std.Error' = standard_error, 'p' = p_value) %>%
    
    #mutate(Value = (Value/Baseline) * 100, Std.Error = (Std.Error/Baseline) * 100) %>%
  
    # Select relevant columns for the specified drug
    #select(SNP, Gene, ch37, ch38, BMD_Site, Value, Std.Error, p, Source, REF, ALT) %>%
    
    # Make sure positions are in numeric terms.
    # Added gsub to remove hidden non-numeric characters from ch38
    mutate(ch38 = sub('.*:', '', ch38)) %>%
    mutate(ch37 = as.numeric(ch37), ch38 = as.numeric(gsub("[^0-9.-]", "", ch38))) %>%

    
    # Add column specifying the significant drug for merge
    mutate(Drug = Drug_Name, .before = BMD_Site) 
    
    
}

atenolol <- parse_data_by_drug(mayo_emmeans, 'SNP_AvsPl_coef', 'SNP_AvsPl_se', 'SNP_AvsPl_pval', 'Atenolol')
nebivolol <- parse_data_by_drug(mayo_emmeans, 'SNP_NvsPl_coef', 'SNP_NvsPl_se', 'SNP_NvsPl_pval', 'Nebivolol')
propranolol_40 <- parse_data_by_drug(mayo_emmeans, 'SNP_P40vsPl_coef', 'SNP_P40vsPl_se', 'SNP_P40vsPl_pval', 'Propranolol40')
propranolol_80 <- parse_data_by_drug(mayo_emmeans, 'SNP_P80vsPl_coef', 'SNP_P80vsPl_se', 'SNP_P80vsPl_pval', 'Propranolol80')

filter_for_significance <- bind_rows(atenolol, nebivolol, propranolol_40, propranolol_80)
```

```{r}
prebuild_means <- filter_for_significance%>%
  
  # select out the predicted emmeans for each copy of the alternate allele
  select(-matches('se')) %>%
  select(SNP, BMD_Site, matches('pred'), matches('pred_se'), Source, Gene, REF, ALT)%>%
  
  # Long to wide, put data in terms of SNP 
  gather(name, EMmean, -SNP, -BMD_Site, -Source, -Gene, -REF, -ALT) %>%
    arrange(SNP)

prebuild_se <- filter_for_significance%>%
  
  # select out the predicted emmeans for each copy of the alternate allele
  select(SNP, BMD_Site, matches('pred_se'), matches('pred_se'), Source, Gene, REF, ALT)%>%
  
  # Long to wide, put data in terms of SNP 
  gather(name, Std.Error, -SNP, -BMD_Site, -Source, -Gene, -REF, -ALT) %>%
    arrange(SNP) %>%
  mutate(name = gsub('*.se','',name))

prebuild <- merge(prebuild_means, prebuild_se) %>%
  # Strip non-numeric characters from the allele
    mutate(Allele = gsub("[^0-9.-]", "", name)) %>%
  
  # Only keep substring after '.' (for P40 and P80) 
    mutate(Allele = gsub('.*\\.','',Allele)) %>%
  
  # Name drug variables so they can be plotted
    mutate(drug_var = case_when(startsWith(name, 'Pl') ~ 'Placebo',
                                startsWith(name, 'A') ~ 'Atenolol',
                                startsWith(name, 'N') ~ 'Nebivolol',
                                startsWith(name, 'P40') ~ 'Propranolol 40',
                                startsWith(name, 'P80') ~ 'Propranolol 80')) %>%
  
  mutate(SNP = case_when(startsWith(SNP, 'rs11124190') ~ 'rs11124190',
                         startsWith(SNP, 'rs1336315111') ~ 'rs1336315111',
                         TRUE ~ SNP)) %>%
  
  # ASSUMES MAYO DATA COUNTS REFERENCE ALLELES, remove this and change
  # labels if Beth says otherwise.
  # Change data to effectively count the number of copies of the alternate
  # allele to better match framingham data.
  
  mutate(Allele = case_when(Allele == '2' ~ 0,
                            Allele == '0' ~ 2,
                            Allele == '1' ~ 1))
  
prebuild
```  


```{r}
make_plots <- function(emmeans_data, bmd_site_list) {
  
  for(i in 1:length(bmd_site_list)) {
    print(bmd_site_list[[i]])
    filtered_df <- emmeans_data %>%
      filter(BMD_Site == bmd_site_list[[i]])
    
    
    for(j in 1:length(unique(filtered_df$SNP))) {
      print(unique(filtered_df$SNP)[[j]])
      plot_interaction_effects <- filtered_df %>%
        filter(SNP == unique(filtered_df$SNP)[[j]])
      
      # Make Plot
       p <-  ggplot(data = plot_interaction_effects, aes(fill = drug_var, x = Allele, y = EMmean)) +
         geom_col(position = 'dodge', color = 'black') +
         scale_y_continuous(labels = scales::percent) +
       geom_errorbar(aes(ymin = EMmean - 1.96 * Std.Error,
                 ymax = EMmean + 1.96* Std.Error),
                 width=.2,position=position_dodge(.9),
                 show.legend = F) +
        scale_fill_manual(values = c('Placebo' = 'grey','Atenolol' = 'light blue', 'Nebivolol' = 'blue', 'Propranolol 40' = 'light green', 'Propranolol 80' = 'dark green')) +
       labs(title = paste(plot_interaction_effects$Gene[[1]], 
                          paste0(plot_interaction_effects$SNP[[1]],'_', plot_interaction_effects$ALT[[1]]),
                          plot_interaction_effects$BMD_Site[[1]], '(Imputed)'),
            x="Number of Copies (Imputed)", y="% Change", fill="Treatment") 
      
        #print(p) 
        ggsave(ggsave(paste0(unique(filtered_df$SNP)[[j]],'_imp_',plot_interaction_effects$BMD_Site[[1]],'.png'),
        plot = p,
        device = png,
        path = '/work/larylab/Griffin_Stuff/Griffin_Mayo/results/plots/'))

    }
  }

}
make_plots(prebuild, list('Femoral Neck', 'Femoral Hip', 'Ultradistal Radius', 'Total Body', 'Lumbar Spine'))
```

```{r}
long_fram <- merge(mayo_snps, long_fram, by.x = 'ch37', by.y = 'POS', all.x = T) %>%
  filter(!is.na(emmean))
cs_fram <- merge(mayo_snps, cs_fram, by.x = 'ch37', by.y = 'POS', all.x = T) %>%
  filter(!is.na(emmean))

bb00 <- cs_fram %>%
  filter(BB == 'No', Allele.y == 0) %>%
  rename(BB_00 = emmean)

cs_fram <- merge(cs_fram, bb00, by = 'SNP') %>% 
  mutate(emmean = emmean - BB_00)
```

```{r}
# Now for Framingham
long_framingham_emmeans_plot <- function(matched_mayo_snps, title, y_variable, suffix){
  for(j in 1:length(unique(matched_mayo_snps$SNP))) {
        print(unique(matched_mayo_snps$SNP)[[j]])
    
        plot_interaction_effects <- matched_mayo_snps %>%
           filter(SNP == unique(matched_mayo_snps$SNP)[[j]])
         
          p <-  ggplot(data = plot_interaction_effects, aes(fill = BB6, x = Allele.y, y = emmean)) +
          geom_col(position = 'dodge', color = 'black') +
          scale_y_continuous(labels = scales::percent) +
        geom_errorbar(aes(ymin = lower.CL, ymax = upper.CL),
                      width=.2,position=position_dodge(.9),
                      show.legend = F) +
         scale_fill_manual(values = c('Yes' = 'light blue', 'No' = 'grey')) +
        labs(title = paste(title, plot_interaction_effects$Gene.x[[1]], 
                            paste0(plot_interaction_effects$SNP[[1]],'_', plot_interaction_effects$ALT[[1]])),
             x="Number of Copies (Imputed)", y= y_variable, fill="Treatment") 
          
          
        ggsave(ggsave(paste0(unique(matched_mayo_snps$SNP)[[j]],suffix,'.png'),
        plot = p,
        device = png,
        path = '/work/larylab/Griffin_Stuff/Griffin_Mayo/results/plots/'))
  }
}

cs_framingham_emmeans_plot <- function(matched_mayo_snps, title, y_variable, suffix){
  for(j in 1:length(unique(matched_mayo_snps$SNP))) {
        print(unique(matched_mayo_snps$SNP)[[j]])
    
        plot_interaction_effects <- matched_mayo_snps %>%
           filter(SNP == unique(matched_mayo_snps$SNP)[[j]])
         
          p <-  ggplot(data = plot_interaction_effects, aes(fill = BB.x, x = Allele.y.x, y = emmean)) +
          geom_col(position = 'dodge', color = 'black') +
        geom_errorbar(aes(ymin = lower.CL.x - BB_00, ymax = upper.CL.x - BB_00),
                      width=.2,position=position_dodge(.9),
                      show.legend = F) +
         scale_fill_manual(values = c('Yes' = 'light blue', 'No' = 'grey')) +
        labs(title = paste(title, plot_interaction_effects$Gene.x.x[[1]], 
                            paste0(plot_interaction_effects$SNP[[1]],'_', plot_interaction_effects$ALT.x[[1]])),
             x="Number of Copies (Imputed)", y= y_variable, fill="Treatment") 
          
          
        ggsave(ggsave(paste0(unique(matched_mayo_snps$SNP)[[j]],suffix,'.png'),
        plot = p,
        device = png,
        path = '/work/larylab/Griffin_Stuff/Griffin_Mayo/results/plots/'))
  }
}

long_framingham_emmeans_plot(long_fram, 'FHS (FN Longitudinal)',
                        '% Change Annualized', '_framingham')
cs_framingham_emmeans_plot(cs_fram, 'FHS (FN Cross Sectional)',
                        'Difference from No BB, Allele: (0/0)', '_cs_framingham')
```
